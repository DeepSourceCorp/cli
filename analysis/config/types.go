package config

import "github.com/deepsourcelabs/cli/analysis/lsp"

///////////////////////////////////////////////////////
// DSConfig is the struct for .deepsource.toml file //
/////////////////////////////////////////////////////

type DSConfig struct {
	Version         int           `mapstructure:"version,omitempty" json:"version" toml:"version"`
	ExcludePatterns []string      `mapstructure:"exclude_patterns,omitempty" json:"exclude_patterns,omitempty" toml:"exclude_patterns"`
	TestPatterns    []string      `mapstructure:"test_patterns,omitempty" json:"test_patterns,omitempty" toml:"test_patterns"`
	Analyzers       []Analyzer    `mapstructure:"analyzers,omitempty" json:"analyzers,omitempty" toml:"analyzers"`
	Transformers    []Transformer `mapstructure:"transformers,omitempty" json:"transformers,omitempty" toml:"transformers"`
}

type Analyzer struct {
	Name                string      `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	RuntimeVersion      string      `mapstructure:"runtime_version,omitempty" json:"runtime_version,omitempty" toml:"runtime_version,omitempty"`
	Enabled             bool        `mapstructure:"enabled,omitempty" json:"enabled" toml:"enabled"`
	DependencyFilePaths []string    `mapstructure:"dependency_file_paths,omitempty" json:"dependency_file_paths,omitempty"`
	Meta                interface{} `mapstructure:"meta,omitempty" json:"meta,omitempty" toml:"meta"`
	Thresholds          interface{} `mapstructure:"thresholds,omitempty" json:"thresholds,omitempty"`
}

type Transformer struct {
	Name    string `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	Enabled bool   `mapstructure:"enabled,omitempty" json:"enabled,omitempty" toml:"enabled"`
}

///////////////////////////////////////
// LSP based Analysis Config Types  //
/////////////////////////////////////

type AnalysisConfig struct {
	Files           []lsp.TextDocumentItem `json:"files"`
	TestFiles       []lsp.TextDocumentItem `json:"test_files"`
	ExcludedFiles   []lsp.TextDocumentItem `json:"excluded_files"`
	ExcludePatterns []string               `json:"exclude_patterns"`
	TestPatterns    []string               `json:"test_patterns"`
	AnalyzerMeta    interface{}            `json:"analyzer_meta"`
}

//////////////////////////////////////
// LSP based Analysis Result Types  //
/////////////////////////////////////

type Namespace struct {
	Key   string  `json:"key"`
	Value float64 `json:"value"`
}

type Metric struct {
	MetricCode string      `json:"metric_code"`
	Namespaces []Namespace `json:"namespaces"`
}

type AnalysisResult struct {
	Issues   []lsp.Diagnostic `json:"issues"`
	Metrics  []Metric         `json:"metrics,omitempty"`
	IsPassed bool             `json:"is_passed"`
	Errors   []Error          `json:"errors"`
	// Errors    []lsp.Diagnostic `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}

////////////////////////////////////
// Default Analysis Result Types  //
///////////////////////////////////

type Error struct {
	HMessage string `json:"hmessage"`
	Level    int    `json:"level"`
}

type Coordinate struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

type Position struct {
	Begin Coordinate `json:"begin"`
	End   Coordinate `json:"end"`
}

type Location struct {
	Path     string   `json:"path"`
	Position Position `json:"position"`
}

type Issue struct {
	Code     string   `json:"issue_code"`
	Title    string   `json:"issue_text"`
	Location Location `json:"location"`
}

type DefaultAnalysisResult struct {
	Issues    []Issue     `json:"issues"`
	Metrics   []Metric    `json:"metrics,omitempty"`
	IsPassed  bool        `json:"is_passed"`
	Errors    []Error     `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}
