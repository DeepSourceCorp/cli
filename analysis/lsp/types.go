package lsp

//////////////////////////////
// Document identity types  //
/////////////////////////////

type DocumentURI string

type TextDocumentItem struct {
	URI        DocumentURI `json:"uri"`
	LanguageID string      `json:"languageID,omitempty"`
	Version    int         `json:"version,omitempty"`
	Text       string      `json:"text,omitempty"`
}

///////////////////////////////
// Diagnostic related types //
/////////////////////////////

type DiagnosticSeverity int

const (
	Error       DiagnosticSeverity = 1
	Warning     DiagnosticSeverity = 2
	Information DiagnosticSeverity = 3
	Hint        DiagnosticSeverity = 4
)

type Position struct {
	Line      int `json:"line"`
	Character int `json:"character"`
}
type Range struct {
	Start Position `json:"start"`
	End   Position `json:"end"`
}

type Location struct {
	URI   string `json:"uri"`
	Range Range  `json:"range"`
}

type DiagnosticRelatedInformation struct {
	Location Location `json:"location"`
	Message  string   `json:"message"`
}

type Diagnostic struct {
	Range    Range              `json:"range"`
	Severity DiagnosticSeverity `json:"severity,omitempty"`
	Code     string             `json:"code,omitempty"`
	Source   string             `json:"source,omitempty"`
	Message  string             `json:"message"`

	/**
	 * An array of related diagnostic information, e.g. when symbol-names within
	 * a scope collide all definitions can be marked via this property.
	 * var a,b
	 * a := 2
	 * Issues in line 1 and 2 are related.
	 */
	RelatedInformation []DiagnosticRelatedInformation `json:"relatedInformation"`
}
