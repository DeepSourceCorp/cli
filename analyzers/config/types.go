package config

type Analysis struct {
	Command string `toml:"command" json:"command" validate:"required"`
}

// TODO: Remove the comment when we start support for Autofix.
// type Autofix struct {
//     Command string `toml:"command" json:"command" analyzertoml:"required"`
// }

type Build struct {
	Builder    string `toml:"builder" json:"builder" validate:"omitempty"`
	Dockerfile string `toml:"dockerfile" json:"dockerfile,omitempty" validate:"omitempty"`
	Script     string `toml:"script" json:"script" validate:"omitempty"`
}

type URL struct {
	Source        string `toml:"source" json:"source" validate:"omitempty,url"`
	Documentation string `toml:"documentation" json:"documentation,omitempty" validate:"omitempty,url"`
	BugTracker    string `toml:"bug_tracker" json:"bug_tracker" validate:"omitempty,url"`
}

type Test struct {
	Command string `toml:"command" json:"command" validate:"omitempty"`
}

// analyzer.toml type
type AnalyzerMetadata struct {
	// Analyzer specific data
	Name                 string            `toml:"name" json:"name" validate:"required"`
	Shortcode            string            `toml:"shortcode" json:"shortcode" validate:"required"` // New
	Description          string            `toml:"description" json:"description" validate:"required"`
	Category             string            `toml:"category" json:"category" validate:"required"`
	Tags                 []string          `toml:"tags" json:"tags,omitempty" validate:"omitempty"`
	URL                  URL               `toml:"urls" json:"urls,omitempty" validate:"omitempty"`
	EnvironmentVariables map[string]string `toml:"environment_variables" json:"environment_variables,omitempty" validate:"omitempty"`

	// Analyzer, Autofix and Transformer config
	Analysis Analysis `toml:"analysis" json:"analysis" validate:"required"`
	// Autofix Autofix `toml:"autofix" json:"autofix" validate:"omitempty"`

	// Build steps needed to build the analyzer
	Build Build `toml:"build" json:"build" validate:"omitempty"`

	// Test command for running the testing workflow
	Test Test `toml:"test" json:"test" validate:"omitempty"`
}

// Analyzer issue description
type AnalyzerIssue struct {
	Shortcode   string `validate:"omitempty"`
	Title       string `toml:"title" json:"title" validate:"required"`
	Description string `toml:"description" json:"description" validate:"required"`
	Category    string `toml:"category" json:"category" validate:"required"`
}
