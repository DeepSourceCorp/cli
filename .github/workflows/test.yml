name: Unit Tests

on:
  push:
    branches: [develop]
  pull_request:
    branches: [develop]

jobs:
  test:
    runs-on: ubuntu-latest
    env:
      NODE_AUTH_TOKEN: ${{secrets.GITHUB_TOKEN}}
      ZEAL_SECRET: ${{secrets.GITHUB_TOKEN}}

    steps:
      - name: Checkout Bifrost
        uses: actions/checkout@v2

      - name: Setting Up Node
        uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          always-auth: true
          registry-url: https://npm.pkg.github.com/
          scope: '@deepsourcelabs'
          cache: 'yarn'

      - name: Install packages
        run: yarn

      - name: Run Test
        run: yarn test --coverage=true --coverageReporters=cobertura

      - name: Install DeepSource CLI
        run: curl https://deepsource.io/cli | sh

      - name: Set DEEPSOURCE_DSN
        run: |
          ./bin/deepsource report --analyzer test-coverage --key javascript --value-file ./coverage/cobertura-coverage.xml
        env:
          DEEPSOURCE_DSN: ${{ secrets.BIFROST_DSN }}
