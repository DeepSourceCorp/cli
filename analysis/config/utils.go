package config

import (
	"log"
	"os"
	"path"
	"path/filepath"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

/* Walks the `CODE_PATH` directory and returns all the files other than the ones present
 * in the .git folder for analysis in the form of a string array */
func readAllFiles(codePath string) ([]string, error) {
	fileCount := 0

	allFiles := make([]string, 0)

	err := filepath.Walk(codePath,
		func(path string, fileInfo os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			fileCount++

			/* Check the following before appending to the list of files:
			 * - Should not be a directory
			 * - The walked file should not be present in .git folder
			 * TODO: Check why fileCount != 1 condition is used here */
			if !fileInfo.IsDir() || !strings.HasPrefix(path, filepath.Join(codePath, ".git")) && fileCount != 1 {
				allFiles = append(allFiles, path)
			}
			return nil
		})
	if err != nil {
		return allFiles, err
	}
	return allFiles, nil
}

// Returns the slice of files matching certain glob patterns
func (r *AnalysisRun) getMatchingFiles(patterns []string) ([]string, error) {
	matchedFiles := make([]string, 0)

	// Return all the files if no exclude_patterns are configured
	if len(patterns) == 0 {
		return matchedFiles, nil
	}

	for _, file := range r.AnalysisFiles {
		for i := range patterns {
			match, err := doublestar.Match(path.Join(r.CodePath, patterns[i]), file)
			if err != nil {
				log.Printf("Failed to match the file %s with pattern %s\n", file, patterns[i])
			}
			if match {
				matchedFiles = append(matchedFiles, file)
				break
			}
		}
	}
	return matchedFiles, nil
}

// Filters the analysis files and removes the files matching the exclude_patterns from them
// TODO: Improve the logic here
func (r *AnalysisRun) filterAnalysisFiles() {
	filteredFiles := []string{}
	for _, file := range r.AnalysisFiles {
		for _, excludedFile := range r.ExcludedFiles {
			if file == excludedFile {
				break
			}
			filteredFiles = append(filteredFiles, file)
		}
	}
	r.AnalysisFiles = filteredFiles
}
