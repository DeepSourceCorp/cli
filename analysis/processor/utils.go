package processor

import (
	"sort"

	"github.com/deepsourcelabs/cli/types"
)

// sortIssuesByFile sorts the issues in an alphabetical order according to the filenames
// where they got reported.
func (p *ProcessAnalysisResults) sortIssuesByFile() {
	sort.Slice(p.AnalysisResult.Issues, func(i, j int) bool {
		return p.AnalysisResult.Issues[i].Location.Path < p.AnalysisResult.Issues[j].Location.Path
	})
}

// Prepare a map with unique filenames as key and the issue range for each file as value
// This is done to ensure fewer loops when processing issues.
type IssueRange struct {
	Filename   string // The file which has the issues
	BeginIndex int    // Array index in report.Issues where the particular issue starts
	EndIndex   int    // Array index in report.Issues where the particular issue ends
}

// GenerateIssueRangeSlice generates an array containing the issue ranges with respect to files
// that helps us to go through them and map them to the files where they got reported instead
// of opening the file for each of them. The generated index looks like this:

// [{analyzer.go 0 0} {autofix_patch.go 1 1} {difftool.go 2 2} {patch.patch 3 5} {proc_skip_cq_test.go 6 6}]

// Here, the first field if filename and the second and third fields are the index range in which the issues reported
// in these files lie in the sorted AnalyzerReport slice.
func createIssueFileRange(report types.AnalysisResult) []IssueRange {
	fileCount := 0 // for 1 file, 0 based indexing
	issuesRange := []IssueRange{}
	prevFilename := report.Issues[0].Location.Path

	issueRange := IssueRange{
		BeginIndex: 0,
		EndIndex:   len(report.Issues) - 1,
		Filename:   prevFilename,
	}
	issuesRange = append(issuesRange, issueRange)

	// Iterating over the issues and creating an array containing issues with index data about
	// the files in which those issues are present
	for i := 1; i < len(report.Issues); i++ {
		issue := report.Issues[i]
		currentFilename := issue.Location.Path

		// TODO: Check when this condition is implied
		if issue.Location.Position.End.Line == -1 {
			issue.Location.Position.End.Line = issue.Location.Position.Begin.Line
		}

		if currentFilename != prevFilename {
			fileCount++

			issueRange = issuesRange[fileCount-1]
			issueRange.EndIndex = i - 1
			issuesRange[fileCount-1] = issueRange

			// Create for the new file
			issueRange = IssueRange{
				Filename:   currentFilename,
				BeginIndex: i,
				EndIndex:   len(report.Issues) - 1,
			}

			issuesRange = append(issuesRange, issueRange)
			prevFilename = currentFilename
		}
	}
	return issuesRange
}

// formatLSPResultsToDefault converts the LSP based analysis results into the default format supported by DeepSource.
func (p *ProcessAnalysisResults) formatLSPResultsToDefault() {
	p.AnalysisResult.IsPassed = p.Report.IsPassed
	p.AnalysisResult.Metrics = append(p.Report.Metrics, p.AnalysisResult.Metrics...)
	p.AnalysisResult.Errors = append(p.Report.Errors, p.AnalysisResult.Errors...)

	// Appending the issues to the default format of Analysis report
	for _, issue := range p.Report.Issues {
		analysisIssue := types.Issue{
			IssueCode: issue.Code,
			IssueText: issue.Message,
			Location: types.Location{
				Path: issue.RelatedInformation[0].Location.URI,
				Position: types.Position{
					Begin: types.Coordinate{
						Line:   issue.Range.Start.Line,
						Column: issue.Range.Start.Character,
					},
					End: types.Coordinate{
						Line:   issue.Range.End.Line,
						Column: issue.Range.End.Character,
					},
				},
			},
		}
		p.AnalysisResult.Issues = append(p.AnalysisResult.Issues, analysisIssue)
	}
}
