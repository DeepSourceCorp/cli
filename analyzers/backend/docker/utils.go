package docker

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

type ErrorLine struct {
	Error       string      `json:"error"`
	ErrorDetail ErrorDetail `json:"errorDetail"`
}

type ErrorDetail struct {
	Message string `json:"message"`
}

type DockerBuildResponse struct {
	Stream string `json:"stream"`
}

func CheckBuildResponse(rd io.Reader, showAllLogs bool) error {
	var lastLine []byte
	count := 0
	var currentStream string

	scanner := bufio.NewScanner(rd)
	for scanner.Scan() {
		lastLine = scanner.Bytes()
		d := &DockerBuildResponse{}
		err := json.Unmarshal(lastLine, d)
		if err != nil {
			return err
		}
		if d.Stream == "" || d.Stream == "\n" || strings.Contains(d.Stream, "--->") || strings.TrimSuffix(d.Stream, "\n") == currentStream {
			continue
		}
		currentStream = strings.TrimSuffix(d.Stream, "\n")
		if showAllLogs {
			fmt.Println(currentStream)
		}
		count++
	}

	errLine := &ErrorLine{}
	json.Unmarshal([]byte(lastLine), errLine)
	if errLine.Error != "" {
		return errors.New(errLine.Error)
	}
	return scanner.Err()
}


