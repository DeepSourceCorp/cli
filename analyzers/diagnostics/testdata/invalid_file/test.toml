name = "2do-Checker"
