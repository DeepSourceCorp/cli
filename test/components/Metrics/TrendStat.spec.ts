import { render } from '@testing-library/vue'
import { TrendStat } from '~/components/Metrics'
import { RouterLinkStub } from '@vue/test-utils'
import { VTooltip } from 'v-tooltip'
import { VueConstructor } from 'vue'
import { cartesian, generateStringProps } from '~/test/utils'
import { StatType } from '~/types/metric'

const injectDirective = (vue: VueConstructor) => vue.directive('tooltip', VTooltip)

test('renders TrendStat with all prop options', () => {
  const baseProps = {
    metric: {
      name: 'Around the world',
      value: 360
    }
  }

  const typeProp = generateStringProps('type', [StatType.metric, StatType.threshold], false)

  cartesian(typeProp).forEach((propCombination) => {
    const props = {
      ...baseProps,
      ...propCombination
    }

    const { html } = render(
      TrendStat,
      {
        props,
        stubs: {
          NuxtLink: RouterLinkStub
        }
      },
      injectDirective
    )

    expect(html()).toMatchSnapshot(JSON.stringify(props))
  })
})
