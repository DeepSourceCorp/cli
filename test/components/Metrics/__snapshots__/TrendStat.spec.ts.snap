// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TrendStat with all prop options: {"metric":{"name":"Around the world","value":360},"type":"metric"} 1`] = `
"<div class=\\"flex flex-col gap-y-2.5\\">
  <div class=\\"flex flex-grow items-end\\">
    <p class=\\"uppercase text-xs text-vanilla-400 tracking-wider font-semibold leading-7\\">
      Around the world
    </p>
  </div>
  <div class=\\"flex items-center gap-x-2 h-6 flex-shrink-0\\"><span class=\\"text-base font-semibold leading-8\\">
      360
    </span>
    <!---->
  </div>
</div>"
`;

exports[`renders TrendStat with all prop options: {"metric":{"name":"Around the world","value":360},"type":"threshold"} 1`] = `
"<div class=\\"flex flex-col gap-y-2.5\\">
  <div class=\\"flex flex-grow items-end\\">
    <p class=\\"uppercase text-xs text-vanilla-400 tracking-wider font-semibold leading-7\\">
      Threshold
    </p>
  </div>
  <div class=\\"flex items-center gap-x-2 h-6 flex-shrink-0\\"><span class=\\"text-base font-semibold leading-8\\">
      360
    </span>
    <!---->
  </div>
</div>"
`;
